File: Resueltos_por_competidor/Juan_Martinez/BuildingRoads.cpp
/*
 * Autor: Juan Martinez
 * Problema: Building Roads
 * Juez online: Cses
 * Veredicto: Accepted
 * Url: https://cses.fi/problemset/task/1666/
 */
#include <bits/stdc++.h>
using namespace std;

vector<int> parent;
vector<int> tam;

void make_set(int node) {
    parent[node] = node;
    tam[node] = 1;
}

int find_set(int node) {
    if(parent[node] == node) return node;
    else return parent[node] = find_set(parent[node]);
}

void union_set(int a, int b) {
    a = find_set(a);
    b = find_set(b);
    if(a != b) {
        if(tam[a] > tam[b]) swap(a,b);
        parent[b] = a;
        tam[a] += tam[b];
    }
} 

int main(){
    int n, m, a, b;
    cin>>n>>m;
    parent.resize(n);
    tam.resize(n);
    for(int i = 0; i < n; i++) make_set(i);  

    for(int i = 0; i < m; i++) {
        cin>>a>>b;
        a--;
        b--;
        union_set(a,b);
    }
    
    set<int> parents;
    for(int i = 0; i < n; i++){
        parents.insert(find_set(i)+1);
        //cout<<find_set(i)+1<<" ";
    }
    //cout<<endl;
    int temp;
    cout<<parents.size()-1<<endl;
    if(parents.size()-1 > 0) {
        for(int i : parents) {
            temp = i;
            break;
        }
        for(int i : parents) {
            if(i != temp) {
                cout<<temp<<" "<<i<<endl;
            }
        }
    }
    return 0;
}

File: Resueltos_por_competidor/Martin_Peña/BuildingRoads.java
/*
* Autor: pgms1103-afk
* Problema: Building Roads
* Juez Online: CSES
* Veredicto: Accepted
* URL: https://cses.fi/problemset/task/1666
*/
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.ArrayList;

public class BuildingRoads {

	static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
	static ArrayList<Integer>[] callesConectadas;
	static boolean[] visitado;
	public static void main(String[] args) throws IOException {
		
		String datos[] = br.readLine().split(" ");
		int numeroCiudades = Integer.parseInt(datos[0]);
		int calles = Integer.parseInt(datos[1]);

		callesConectadas = new ArrayList[numeroCiudades+1];	
		
		for (int i = 0; i <= numeroCiudades; i++) {
			callesConectadas[i] = new ArrayList<>();
		}
		
		for (int i = 0; i < calles; i++) {
			String datos2[] = br.readLine().split(" ");
			int u = (Integer.parseInt(datos2[0]));
			int v = (Integer.parseInt(datos2[1]));
			
			callesConectadas[u].add(v);
			callesConectadas[v].add(u);
		}
		
		visitado = new boolean[numeroCiudades + 1];
		ArrayList<Integer> representantes = new ArrayList<>();
		int nuevasCarreteras = 0;
		for (int i = 1; i <= numeroCiudades; i++) {
			if(!visitado[i]) {
				nuevasCarreteras++;
				representantes.add(i);
				dfs(i);	
			}
		}
		
		System.out.println(nuevasCarreteras-1);
		if(nuevasCarreteras-1 != 0) {
			for (int i = 0; i < representantes.size()-1; i++) {
				int primero = representantes.get(i);
				int segundo = representantes.get(i+1);
				System.out.println(primero + " " + segundo);
			}
		}
		
	}
	
	static void dfs(int nodo) {
		visitado[nodo] = true;
		for (int vecino :  callesConectadas[nodo]) {
			if(!visitado[vecino]) {
				dfs(vecino);
			}
		}
		
	}

}
